File: lc/coreprep/implementation/patterns_reference_implementation/top_k_elements/KClosestNumbersTwoPointers.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KClosestNumbersTwoPointers {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();
        int x = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        var res = kClosestNumbers(a, k, x);
        for (int r : res) {
          System.out.printf("%d ", r);
        }
        System.out.println();
      }
    }
  }

  // O(logn + k) / O(1)
  public static List<Integer> kClosestNumbers(int[] a, int k, int x) {
    int n = a.length;
    int closestIdx = binarySearch(a, 0, n - 1, x);

    List<Integer> res = new LinkedList<>();
    int left = closestIdx, right = closestIdx + 1;

    for (int i = 0; i < k; i++) {
      if (left >= 0 && right <= n - 1) {
        if (Math.abs(a[left] - x) < Math.abs(a[right] - x)) {
          res.add(0, a[left]);
          left--;
        } else {
          res.add(a[right]);
          right++;
        }
      } else if (left >= 0) {
        res.add(0, a[left]);
        left--;
      } else {
        res.add(a[right]);
        right++;
      }
    }

    return res;
  }

  private static int binarySearch(int[] a, int low, int high, int k) {
    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] < k) {
        low = mid + 1;
      } else if (a[mid] > k) {
        high = mid - 1;
      } else {
        return mid;
      }
    }

    if (low > 0) {
      low--;
    }
    return low;
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/top_k_elements_1/k_closest_numbers_two_pointers.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

int binary_search(const vector<int> &a, int low, int high, int elem) {
  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] == elem) {
      return mid;
    } else if (a[mid] < elem) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  if (low > 0) {
    low--;
  }

  return low;
}

// O(logn + k) / O(1)
vector<int> k_closest_numbers(const vector<int> &a, int k, int x) {
  int n = a.size();
  int xpos = binary_search(a, 0, n - 1, x);
  deque<int> q;

  int left = xpos, right = xpos + 1;
  for (int i = 0; i < k; i++) {
    if (left >= 0 && right <= n - 1) {
      if (abs(a[left] - x) < abs(a[right] - x)) {
        q.push_front(a[left]);
        left--;
      } else {
        q.push_back(a[right]);
        right++;
      }
    } else if (left >= 0) {
      q.push_front(a[left]);
      left--;
    } else {
      q.push_back(a[right]);
      right++;
    }
  }

  vector<int> res(q.begin(), q.end());
  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k, x;
  cin >> tt;

  while (tt--) {
    cin >> n >> k >> x;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    auto res = k_closest_numbers(a, k, x);
    for (int r : res) {
      cout << r << " ";
    }
    cout << "\n";
  }

  return 0;
}